AmplifyPlugins/Core/AWSPluginsCore/Model/Decorator/AuthRuleDecorator.swift

```diff
@@ -0,0 +1,106 @@
+//
+// Copyright 2018-2020 Amazon.com,
+// Inc. or its affiliates. All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+import Amplify
+
+// https://github.com/aws-amplify/amplify-cli/issues/4182
+// If the owner authorization does not contain `.read` operation, then others can read owners data,
+// do not have to apply any owner field to the subscription document input
+/// Decorate the document with auth related fields such as the selection set and owner input to subscription documents
+public struct AuthRuleDecorator: ModelBasedGraphQLDocumentDecorator {
+
+    private let subscriptionType: GraphQLSubscriptionType?
+    private let ownerId: String?
+
+    public init(subscriptionType: GraphQLSubscriptionType? = nil, ownerId: String? = nil) {
+        self.subscriptionType = subscriptionType
```

Comment:
Is this optional because we are using this decorator for both subscriptions out mutations?  If so, maybe it would be better to introduce a new type/enum which describes what we are decorating?

Reply:
yeah it's used for both, adding enum like
```
public enum AuthRuleDecoratorInput {
    public typealias OwnerId = String
    case subscription(GraphQLSubscriptionType, OwnerId?)
    case mutation
}
```
to differentiate the two